src/ScheduleParam.h

```
@@ -0,0 +1,254 @@
+        } else if (new_value_string == "inline") {
+            set(LoopLevel::inlined());
+        } else {
+            user_error << "Unable to parse " << name() << ": " << new_value_string;
+        }
+    }
+
+    template <typename T2 = T,
+              typename std::enable_if<std::is_same<T2, bool>::value>::type * = nullptr>
+    NO_INLINE void set_from_string_impl(const std::string &new_value_string) {
+        if (new_value_string == "true") {
+            set(true);
+        } else if (new_value_string == "false") {
+            set(false);
+        } else {
+            user_error << "Unable to parse " << name() << ": " << new_value_string;
+        }
+    }
+
+    template <typename T2 = T,
+              typename std::enable_if<std::is_arithmetic<T2>::value && !std::is_same<T2, bool>::value>::type * = nullptr>
+    NO_INLINE void set_from_string_impl(const std::string &new_value_string) {
+        std::istringstream iss(new_value_string);
+        T t;
+        iss >> t;
+        user_assert(!iss.fail() && iss.get() == EOF) << "Unable to parse " << name() << ": " << new_value_string;
+        set(t);
+    }
+
+    void set_from_string(const std::string &new_value_string) override {
+        set_from_string_impl(new_value_string);
+    }
+
+public:
+    using type = T;
+
+    ScheduleParam() : Internal::ScheduleParamBase(get_param_type()) {}
+
+    explicit ScheduleParam(const std::string &name) : Internal::ScheduleParamBase(get_param_type(), name) {}
+
+    ScheduleParam(const std::string &name, const T &value) : Internal::ScheduleParamBase(get_param_type(), name) {
+        set(value);
+    }
+
+    ScheduleParam(const std::string &name, const std::string &value) : Internal::ScheduleParamBase(get_param_type(), name) {
+        set_from_string(value);
+    }
+
+#define HALIDE_SCHEDULE_PARAM_TYPED_SETTER(TYPE) \
+    void set(const TYPE &new_value) override { typed_setter_impl<TYPE>(new_value, #TYPE); }
+
+    HALIDE_SCHEDULE_PARAM_TYPED_SETTER(bool)
```

It seems to be so that ScheduleParam<int>::set(int) doesn't accept and silently truncate ScheduleParam<int>::set(float). Is that accurate?